src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -0,0 +1,769 @@
+        private async Task ClearRowsAsync(string error, bool acquireLock)
+        {
+            if (acquireLock)
+            {
+                this._logger.LogError(error);
+                await this._rowsLock.WaitAsync();
+            }
+
+            this._leaseRenewalCount = 0;
+            this._state = State.CheckingForChanges;
+            this._rows = new List<IReadOnlyDictionary<string, string>>();
+            this._rowsLock.Release();
+        }
+
+        /// <summary>
+        /// Releases the leases held on "_rows".
+        /// </summary>
+        /// <returns></returns>
+        private async Task ReleaseLeasesAsync(CancellationToken token)
+        {
+            // Don't want to change the "_rows" while another thread is attempting to renew leases on them.
+            await this._rowsLock.WaitAsync(token);
+            long newLastSyncVersion = this.RecomputeLastSyncVersion();
+
+            try
+            {
+                using (var connection = new SqlConnection(this._connectionString))
+                {
+                    await connection.OpenAsync(token);
+                    using (SqlTransaction transaction = connection.BeginTransaction(System.Data.IsolationLevel.RepeatableRead))
+                    {
+                        try
+                        {
+                            // Release the leases held on "_rows".
+                            using (SqlCommand releaseLeasesCommand = this.BuildReleaseLeasesCommand(connection, transaction))
+                            {
+                                await releaseLeasesCommand.ExecuteNonQueryAsync(token);
+                            }
+
+                            // Update the global state table if we have processed all changes with ChangeVersion <= newLastSyncVersion,
+                            // and clean up the worker table to remove all rows with ChangeVersion <= newLastSyncVersion.
+                            using (SqlCommand updateTablesPostInvocationCommand = this.BuildUpdateTablesPostInvocation(connection, transaction, newLastSyncVersion))
+                            {
+                                await updateTablesPostInvocationCommand.ExecuteNonQueryAsync(token);
+                            }
+
+                            transaction.Commit();
+                        }
+                        catch (Exception ex)
+                        {
+                            this._logger.LogError("Commit Exception Type: {0}", ex.GetType());
+                            this._logger.LogError("  Message: {0}", ex.Message);
+                            try
+                            {
+                                transaction.Rollback();
+                            }
+                            catch (Exception ex2)
+                            {
+                                this._logger.LogError("Rollback Exception Type: {0}", ex2.GetType());
```

Comment:
Follow Charles's above suggestion for all logging statements (something like Exception releasing leases on rows for table...)